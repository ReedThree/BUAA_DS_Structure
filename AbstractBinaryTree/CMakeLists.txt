add_library(AbstractBinaryTree AbstractBinaryTree.c)
target_link_libraries(AbstractBinaryTree PRIVATE structure_compiler_flags Trace)
target_include_directories(AbstractBinaryTree INTERFACE ${CMAKE_CURRENT_SOURCE_DIR})