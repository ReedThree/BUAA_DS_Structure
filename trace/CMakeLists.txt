add_library(Trace trace.c _trace_ArrayList.c)
target_link_libraries(Trace PRIVATE structure_compiler_flags)
target_include_directories(Trace INTERFACE ${CMAKE_CURRENT_SOURCE_DIR})