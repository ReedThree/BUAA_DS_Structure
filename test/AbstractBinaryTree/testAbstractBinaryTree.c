#include <stdio.h>

#include "AbstractBinaryTree.h"
#include "trace.h"

void f(struct BinaryTree_Node *node);
void printPath(struct BinaryTree_Node **path);
void checkPath(struct BinaryTree_Node *node);
void checkRoot(struct BinaryTree_Node *node);

static struct BinaryTree_Node *testHead = NULL;

int main(void) {
    traceInit();

    char input[] = "ABD^^EJ^^^CF^I^^G^^";

    struct BinaryTree *tree = BinaryTree_fromPattern(input);
    testHead = tree->root;
    printf("1:\n");
    printf("%zu\n", BinaryTree_getNodeCount(tree));
    printf("2:\n");
    printf("%zu\n", BinaryTree_getHeight(tree));
    printf("3:\n");
    BinaryTree_forEach(tree, 3, f);
    putchar('\n');
    printf("4:\n");
    BinaryTree_forEach(tree, 3, checkPath);
    printf("5:\n");
    BinaryTree_forEach(tree, 3, checkRoot);
    printf("6:\n");
    BinaryTree_forEach(tree, 0, f);
    putchar('\n');
    printf("7:\n");
    BinaryTree_forEach(tree, 1, f);
    putchar('\n');
    printf("8:\n");
    BinaryTree_forEach(tree, 2, f);
    putchar('\n');
    printf("9:\n");
    BinaryTree_forEach(tree, 3, f);
    putchar('\n');

    struct BinaryTree *cloned = BinaryTree_clone(tree);
    printf("10:\n");
    BinaryTree_forEach(tree, 3, f);
    putchar('\n');

    struct BinaryTree *test = BinaryTree_recover("HEIBADFCG", 0, "ABEHICDFG");

    printf("11:\n");
    BinaryTree_forEach(test, 3, f);
    putchar('\n');

    BinaryTree_destroyTree(test);
    test = BinaryTree_recover("HEIBADFCG", 1, "HIEBFDGCA");

    printf("12:\n");
    BinaryTree_forEach(test, 3, f);
    putchar('\n');

    BinaryTree_destroyTree(tree);
    BinaryTree_destroyTree(cloned);
    BinaryTree_destroyTree(test);
    traceFree();
    return 0;
}

BinaryTree_Data *BinaryTree_copyData(BinaryTree_Data *source) {
    BinaryTree_Data *result =
        (BinaryTree_Data *)_malloc(sizeof(BinaryTree_Data));
    *result = *source;
    return result;
}
void BinaryTree_destroyData(BinaryTree_Data *target) { _free(target); }

void f(struct BinaryTree_Node *node) { printf("%c", *(node->data)); }

void printPath(struct BinaryTree_Node **path) {

    size_t i = 0;
    struct BinaryTree_Node *current = path[i];

    while (current != NULL) {
        printf("%c ", *(current->data));
        i++;
        current = path[i];
    }
    putchar('\n');
}

void checkPath(struct BinaryTree_Node *node) {
    struct BinaryTree_Node **path = BinaryTree_getPath(testHead, node);
    if (path != NULL) {
        printf("%c: ", *(node->data));
        printPath(path);
        _free(path);
    }
}
void checkRoot(struct BinaryTree_Node *node) {
    struct BinaryTree_Node *root = BinaryTree_getRoot(node);
    if (root == testHead) {
        printf("Success!Rock and stone!\n");
    } else {
        printf("Failed!\n");
    }
}
