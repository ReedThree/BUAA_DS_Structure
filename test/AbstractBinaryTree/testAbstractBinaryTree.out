1:
9
2:
4
3:
ABCDEFGJI
4:
B: A B 
C: A C 
D: A B D 
E: A B E 
F: A C F 
G: A C G 
J: A B E J 
I: A C F I 
5:
Success!Rock and stone!
Success!Rock and stone!
Success!Rock and stone!
Success!Rock and stone!
Success!Rock and stone!
Success!Rock and stone!
Success!Rock and stone!
Success!Rock and stone!
Success!Rock and stone!
6:
ABDEJCFIG
7:
DBJEAFICG
8:
DJEBIFGCA
9:
ABCDEFGJI
10:
ABCDEFGJI
11:
ABCEDGHIF
12:
ABCEDGHIF
13:
A
B C 
D E F G 
^ ^ J ^ ^ I ^ ^ 
^ ^ ^ ^ 
14:
graph G {
node [shape = circle;];
0 [label = "65";];
1 [label = "66";];
2 [label = "67";];
3 [label = "68";];
4 [label = "69";];
5 [label = "70";];
6 [label = "71";];
7 [label = "74";];
8 [label = "73";];
0 -- 1;
0 -- 2;
1 -- 3;
1 -- 4;
2 -- 5;
2 -- 6;
4 -- 7;
5 -- 8;
}
